mino/minogrpc/routing.go

```diff
@@ -0,0 +1,201 @@
+package minogrpc
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/binary"
+	"fmt"
+	math "math"
+	"math/rand"
+	"sort"
+	"strings"
+
+	"go.dedis.ch/fabric"
+	"go.dedis.ch/fabric/mino"
+	"golang.org/x/xerrors"
+)
+
+// TreeRouting holds the routing tree of a network. It allows each node of the
+// tree to know which child it should contact in order to relay a message that
+// is in it sub-tree.
+type TreeRouting struct {
+	root         *treeNode
+	me           *treeNode
+	routingNodes map[string]*treeNode
+}
+
+// NewTreeRouting creates the network tree in a deterministic manner based on
+// the list of addresses. Please be careful that the provided list of addresses
+// is shuffled, which could affect subsequent use of this list, especially
+// outside the function's scope.
+func NewTreeRouting(addrs []mino.Address, addr mino.Address,
+	treeHeight int) (*TreeRouting, error) {
+
+	addrsStr := make([]string, len(addrs))
+	for i, addr := range addrs {
+		addrsStr[i] = addr.String()
+	}
+	sort.Strings(addrsStr)
+
+	// We will use the hash of the addresses to set the random seed.
+	hash := sha256.New()
+	for _, addr := range addrsStr {
+		_, err := hash.Write([]byte(addr))
+		if err != nil {
+			fabric.Logger.Fatal().Msgf("failed to write hash: %v", err)
+		}
+	}
+
+	hashBuf := bytes.NewBuffer(hash.Sum(nil))
+	hashInt, err := binary.ReadVarint(hashBuf)
```

Comment:
It's easier to implement this in a different language if it is defined properly like `binary.LittleEndian.ReadUInt64`. Because you have to look for the Varint way of reading.

Reply:
As you're using sha256, you know the size.

Reply:
something like `binary.LittleEndian.Uint64(hash.Sum(nil))` ?

Reply:
yep that one